calculator_test.go

```diff
@@ -56,141 +49,174 @@ func TestAddRandom(t *testing.T) {
-			t.Errorf("%s, Add(%v) = %f, want: %f", name, a, got, want)
+			t.Errorf("%s; Add() = %f; want: %f", name, got, want)
 		}
 	}
 }
 
-func TestSubtract(t *testing.T) {
+func TestCalculator(t *testing.T) {
 	t.Parallel()
 
-	tt := []struct {
-		name        string
-		nums        []float64
-		want        float64
-		expectedErr bool
-	}{
-		{name: "Substract two positive numbers", nums: []float64{4, 2}, want: 2, expectedErr: false},
-		{name: "Substract two negative numbers", nums: []float64{-4.5, -5.5}, want: 1, expectedErr: false},
-		{name: "Substract from zero", nums: []float64{0, 5.45}, want: -5.45, expectedErr: false},
-		{name: "Substract multiple numbers", nums: []float64{20, 5.5, 4, 3, 1.25}, want: 6.25, expectedErr: false},
-		{name: "Substract incorrect number of params", nums: []float64{4}, want: 0, expectedErr: true},
-		{name: "Substract incorrect number of params", nums: []float64{}, want: 0, expectedErr: true},
+	tt := []testCase{
+		{name: "Add two positive numbers", a: 1, b: 2, want: 3},
+		{name: "Add two negative numbers", a: -4.5, b: -5.5, want: -10},
+		{name: "Add three negative numbers", a: -4.5, b: -5.5, c: []float64{-4.5}, want: -14.5},
+		{name: "Add multiple numbers", a: 2, b: 3.4, c: []float64{-4.5, -5.5, 10, 2.5}, want: 7.9},
+
+		{name: "Multiply three positive", a: 3, b: 2, c: []float64{2}, want: 12},
+		{name: "Multiply two positive one negative", a: 3, b: 2, c: []float64{-2}, want: -12},
+		{name: "Multiply one positive two negative", a: -3, b: 2, c: []float64{-2.5}, want: 15},
+		{name: "Multiply multiple numbers", a: 3, b: 2, c: []float64{2, -1}, want: -12},
+		{name: "Multiply two negative numbers", a: 2.5, b: 1.5, c: []float64{-4, -5}, want: 75},
+		{name: "Multiply by zero", a: 3, b: 4, c: []float64{0, -4}, want: 0},
+
+		{name: "Subtract two positive numbers", a: 4, b: 2, want: 2},
+		{name: "Subtract two negative numbers", a: -4.5, b: -5.5, want: 1},
+		{name: "Subtract from zero", a: 0, b: 5.45, want: -5.45},
+		{name: "Subtract multiple numbers", a: 10, b: 3, c: []float64{2, 1, 4, 3}, want: -3},
+		{name: "Subtract multiple numbers with negative", a: 10.5, b: 3, c: []float64{20, 5.5, 4, -3, 1.25}, want: -20.25},
 	}
 
 	for _, tc := range tt {
-		got, err := calculator.Subtract(tc.nums...)
-
-		if err != nil && !tc.expectedErr {
-			t.Errorf("%s, Substract(%v) should return an error", tc.name, tc.nums)
-		}
-
-		if tc.want != got && !tc.expectedErr {
-			t.Errorf("%s, Substract(%v) = %f, want %f", tc.name, tc.nums, got, tc.want)
-		}
+		t.Run(tc.name, func(t *testing.T) {
+
+			// we need to decide which calculator func to run. I took approach
+			// of not adding a new field in the test struct (to differentiate
+			// function test data) but to keep test name consistent -
+			// always start with the calc function name. This way I can
+			// parse name, extract the function name and use it in the switch
+			// statement.
+			fname := strings.Split(tc.name, " ")[0]
+			var got float64
+
+			switch fname {
```

Comment:
Yes, you could do this, but it clearly wouldn't scale very well, and is unnecessary in a language with first-class functions! We could instead add a field to the struct that stores the function to be called (for example `calculator.Add`).

Reply:
That's a brilliant idea. Thank you, John! I have applied your suggestion. I was wondering if, instead of passing a func as a field in a struct would be correct to run the function and have, for example ```got``` field in the test struct, sth like this:

current (after applied code review)
```go
tt := []struct {
    name string
    fn   func(a, b float64, c ...float64) float64
    a    float64
    b    float64
    c    []float64
    want float64
}{
    {name: "Add two positive numbers", fn: calculator.Add, a: 1, b: 2, want: 3},
    {name: "Multiply three positive", fn: calculator.Multiply, a: 3, b: 2, c: []float64{2}, want: 12},
    . . .
}

for _, tc := range tt {
    t.Run(tc.name, func(t *testing.T) {
        got := tc.fn(tc.a, tc.b, tc.c...)
	if got != tc.want {
            t.Errorf("%s, got: %v, want: %v", tc.name, got, tc.want)
	}
     })
}
```

another solution I have in mind (limited to functions that return only one value)
```go
tt := []struct{
    name string
    got float64
    want float64
}{
    {name: "Add two positive numbers", got: calculator.Add(1, 2), want: 3}
    . . . 
}

for _, tc := range tt {
    t.Run(tc.name, func(t *testing.T) {
	if tc.got != tc.want {
            t.Errorf("%s, got: %v, want: %v", tc.name, tc.got, tc.want)
	}
     })
}
```

Reply:
Yes, both of those are fine! It's cool to be able to use functions as variables (or struct fields), so it's good to know how to do that, but choose whichever way you prefer!